🚀 3단계 - 경로 조회 기능

wonseok5893 opened the pull request:
2단계 피드백 감사합니다!
다음과 같이 진행했습니다.
- 경로 검색 기능 인수테스트 작성
- 최단거리 테스트
- 최단거리 서비스 구현
- refactor
- 예외 인수테스트 작성
- 구현 및 테스트 통과
- refactor

Commit:
2단계 피드백 반영
- import 제거
- remove도 전략 패턴 풀어보기.
- Holder를 통한 책임 분배
- 중간 제거 로직상 버그 수정 -> 삭제한 구간 idx에 구간 추가
- 도메인 객체 불변 리스트로 반영

Commit:
step3 - 경로 조회 인수테스트 작성

Commit:
step3 - 최단 거리 조회 테스트

Commit:
step3 - 최단거리 구현 및 실패 인수테스트 작성

Commit:
step3 - 실패 인수테스트 구현 및 refactor

catsbi commented on src/main/java/nextstep/subway/linesection/LineSections.java:
```diff
@@ -57,33 +60,24 @@ public boolean containsStation(Station station) {
-        if (deleteStation.equals(getLastStation())) {
-            sections.remove(getLastSection());
-            return;
+        for (LineSectionRemover remover : LineSectionRemoverHolder.getContext()) {
+            if (remover.remove(this, deleteStation))
+                return;
```
반복문내에 분기문이 있다보니  indent가 3이 되버렸어요. Stream API를 이용해서 분기문을 없앨 수 있지 않을까요?
```java
LineSectionRemoverHolder.getContext().stream().filter(LineSectionRemover::support) ... 
```

이후에 찾아서 수행하든 있을경우 컨슘을 하든 할 수 있지 않을까요

catsbi commented on src/main/java/nextstep/subway/linesection/append/DownStationLineSectionAppender.java:
```diff
@@ -10,13 +10,12 @@ public class DownStationLineSectionAppender implements LineSectionAppender {
-        List<LineSection> sectionList = sections.getSections();
         if (sections.containsStation(addSection.getDownStation())) {
-            LineSection section = sectionList.get(idx);
+            LineSection section = sections.getSections().get(idx);
```
내부값을 꺼내서(getSections), 연산을 수행(get(idx))하는건OOP패러다임인 객체간에 요청하고 응답한다와는 거리가 먼 것 같아요. 어떻게 바꿀 수 있을까요?

catsbi commented on src/main/java/nextstep/subway/linesection/append/DownStationLineSectionAppender.java:
```diff
@@ -10,13 +10,12 @@ public class DownStationLineSectionAppender implements LineSectionAppender {
-            sectionList.add(idx, LineSection.of(addSection.getLine(), section.getUpStation(), addSection.getUpStation(), section.getDistance() - addSection.getDistance()));
-            sectionList.add(idx + 1, LineSection.of(addSection.getLine(), addSection.getUpStation(), section.getDownStation(), addSection.getDistance()));
-            sectionList.remove(section);
+            sections.addToIndex(idx, LineSection.of(addSection.getLine(), section.getUpStation(), addSection.getUpStation(), section.getDistance() - addSection.getDistance()));
+            sections.addToIndex(idx + 1, LineSection.of(addSection.getLine(), addSection.getUpStation(), section.getDownStation(), addSection.getDistance()));
+            sections.removeSection(section);
```
코드의 길이가 좀 더 길어지더라도 메서도 호출부에 인수를 생성하는 로직을 넣는것보다 가독성 측면에서 유리할 수 있습니다

wonseok5893 replied:
@catsbi 어떤 의미인지 정확하게 이해하지는 못했지만, 
Appender ->  LineSections 추가해야할 관련 정보를 보내고, LineSections에서 LineSection을 생성하여 추가하게끔 수정하도록하겠습니다.

catsbi commented on src/main/java/nextstep/subway/linesection/append/LineSectionAppenderHolder.java:
```diff
@@ -0,0 +1,15 @@
+
+import java.util.List;
+
+public class LineSectionAppenderHolder {
```
극단적으로 다음과 같이 코드를 작성하면 어떻게 될까요? 
```java
for(int i = 0; i < Integer.MAX_VALUE; i++) {
    new LineSectionAppenderHolder()
}
```

wonseok5893 replied:
싱글톤으로 관리하기 위해 빈으로 등록해서 사용하도록 하겠습니다ㅎㅎ

wonseok5893 replied:
다시생각해보니, LineSectionAppenderHolder 빈으로 관리할 시에, 불편한 느낌이 드는 부분
1. LineSections는 도메인 관리 대상이기 때문에 해당 Appender를 도메인에 DI하는게 아니다라고 생각이 들어요.
2. 1번을 피하기 위해 빈으로 관리 + 정적으로 관리하기에 DI개념과 조금 어긋난게 아닌가 생각이 들어요.
결론 빈 관리가 아닌 싱글톤을 구현하여 사용하는 방식으로 해야할것 같습니다.

catsbi commented on src/main/java/nextstep/subway/linesection/remove/LineSectionRemover.java:
```diff
@@ -0,0 +1,14 @@
+import nextstep.subway.station.Station;
+
+public interface LineSectionRemover {
+    /**
+     *
+     * @param sections
+     * @param deleteStation
+     * @return removed : boolean
+     */
```
javadocs를 작성하고자 한다면 내용을 채워주면 좋을 것 같네요. 
https://catsbi.oopy.io/1b439cd5-58aa-4cb1-a715-2bfa66afc72a#bb57419c-9dcc-478c-8e54-e067d5bef112

wonseok5893 replied:
참고 문서 감사합니다. 
해당 부분은 다른 미션에 집중하기 위해서 미션 완료 후에 한번 반영해보도록 하겠습니다!

catsbi commented on src/main/java/nextstep/subway/linesection/remove/MiddleStationRemover.java:
```diff
@@ -0,0 +1,18 @@
+public class MiddleStationRemover implements LineSectionRemover {
+    @Override
+    public boolean remove(LineSections sections, Station deleteStation) {
+        LineSection beforeSection = sections.findSectionByCondition(section -> section.getDownStation().equals(deleteStation));
+        LineSection nextSection = sections.findSectionByCondition(section -> section.getUpStation().equals(deleteStation));
+        int idx = sections.getIndex(beforeSection);
+        sections.removeSection(beforeSection);
+        sections.removeSection(nextSection);
+        sections.addToIndex(idx, LineSection.of(beforeSection.getLine(), beforeSection.getUpStation(), nextSection.getDownStation(), beforeSection.getDistance() + nextSection.getDistance()));
+        return true;
```
코드의 가독성을 더 높혀볼 수 없을까요?

catsbi commented on src/main/java/nextstep/subway/path/PathService.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    private static void validatePath(ShortestPath<Station> shortestPath) {
+        if (!shortestPath.isFound())
+            throw new BadRequestException("the path is not found.");
```
한 줄짜리 분기문도 블럭`{}` 생략은 지양해주세요!

catsbi commented on src/main/java/nextstep/subway/path/PathService.java:
```diff
@@ -0,0 +1,41 @@
+                stationService.getStation(source),
+                stationService.getStation(target));
+        validatePath(shortestPath);
+        var stations = shortestPath.getVertexList()
+                .stream()
+                .map(e -> stationConverter.convert(e))
+                .collect(Collectors.toList());
```
저는 개인적으로 코틀린이 아닌 자바에서 var나 val을 사용하는걸 크게 권장하지는 않아요. 
컴파일러가 뛰어나서 타입추론을 한다지만 개발자 입장에서 타입추론이 쉽지 않거든요.

wonseok5893 replied:
공감합니다ㅎㅎ

wonseok5893 replied:
유연해 진다는 장점도 있을 것 같아요. 
예를들어, 메소드 단위로 테스트코드 작성시에 return 타입이 변경되는 경우, 유연하게 var or val를 사용해서 편하게 타입추론을 통해 수정을 하지 않아도 되었던 장점이 있었던 경험도 있어요 ㅎㅎ

catsbi commented on src/main/java/nextstep/subway/path/ShortestPath.java:
```diff
@@ -0,0 +1,34 @@
+    private final Integer shortestDistance;
+    private final boolean isFound;
+
+    public ShortestPath(WeightedMultigraph<V, DefaultWeightedEdge> graphPath, V source, V target) {
+        var shortestPath = new DijkstraShortestPath<>(graphPath);
+        var path = shortestPath.getPath(source, target);
+        isFound = path != null;
+        vertexList = isFound ? path.getVertexList() : List.of();
+        shortestDistance = isFound ? (int) shortestPath.getPathWeight(source, target) : 0;
+    }
```
생성자에서 필드 할당이 아닌 의미있는 비즈니스 로직을 작성하는건 권장드리지 않습니다. 비용측정이 힘들 뿐더러 메서드 시그니처를 보고 우리는 고비용이 들꺼라 유추할 수 없기 때문이죠. 협업 혹은 내가 해당 내용을 잊을 경우 어떤 사이드 이펙트가 생길 지 알 수 없습니다

wonseok5893 replied:
의견 감사합니다! 피드백 반영해보도록 하겠습니다.

catsbi reviewed (changes requested):
wonseok5893님 막판 스퍼트 열심히 해주시네요 💯 
경로탐색 관련하여 피드백 남긴부분들이 있으니 확인해주시고 다시 리뷰요청 부탁드려요!

Commit:
step3 - 피드백 반영

1. Appender, Remover indent 축소
2. Appender, Remover 싱글톤 처리
3. val, var 제거
4. 생성자에 비즈니스로직 제거
5. 기타 피드백 반영

wonseok5893 commented:
피드백 반영했습니다. 감사합니다 ㅎㅎ

catsbi commented on src/main/java/nextstep/subway/linesection/LineSections.java:
```diff
@@ -50,6 +51,12 @@ public int getIndex(LineSection section) {
     }
 
+    public LineSection getSectionByIndex(int idx) {
+        if (idx < 0 || sections.size() <= idx)
+            throw new BadRequestException("idx must be between 0 and sections size");
```
블럭`{}` 생략은 지양해주세요 ㅠㅠ

catsbi commented on src/main/java/nextstep/subway/linesection/LineSections.java:
```diff
@@ -137,8 +146,8 @@ public void addInFirst(LineSection addSection) {
 
-    public void addToIndex(int idx, LineSection addSection) {
-        sections.add(idx, addSection);
+    public void addToIndex(int idx, Line line, Station upStation, Station downStation, Integer distance) {
```
파라미터의 갯수가 2개를 넘어가네요!
이를 제가 기존에 드렸던 피드백은 메서드 호출하면서 매개변수를 작성하는 공간에 다른 메서드를 호출하는 식으로 작성하면 디버깅이 힘드니까 매개변수를 만들기 위한 메서드 호출(or 객체 생성을 위한 생성자나 정적 팩토리 호출)로직을 분리하라는 의미였습니다 ㅎㅎ 아니면 추가용 빌더객체를 구간일급컬렉션 내부적으로 호출할수도 있구요.

catsbi reviewed (approved):
원석님 마지막까지 피드백 반영 잘 해주셨어요. 
미션 진행하느라 고생 많으셨습니다. 피드백 남긴부분과 코멘트만 확인해주시면 좋을 것 같아요. 
이만 merge하도록 하겠습니다!
